### PR TITLE
uplifting to chef14

### PR DESCRIPTION
**Pull request by som3guy**
As a consumer of:
cerner_splunk

I want to:
Uplift the cookbook to be compatible with chef-client version 14

So that:
We can continue to use this cookbook on supported chef products

We'll know we are done when:
This PR is merged

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ language: ruby
 cache: bundler
 sudo: false
 rvm:
-- 2.3.1
+- 2.4.1
```
Comment (nmattam) [inline .travis.yml]: Looks like Chef 14 comes with ruby 2.5 https://docs.chef.io/release_notes.html#what-s-new-in-14-0. Can we go to 2.5?

File: Gemfile [hunk 2]
```diff
@@ -7,7 +7,7 @@ rubocop_version = '= 0.48.1'
 chef_vault_version = '> 3.0'
 
 # https://github.com/cerner/cerner_splunk/issues/142
-chef_version = '= 12.18.31'
+chef_version = '~> 14'
```
Comment (nmattam) [inline Gemfile]: If this PR fixes the issue mentioned on line 9, we can remove that comment.

**Comment by som3guy** [comment]
I have updated the Vagrantfile and was able get successful converges for both sections int the [vagrant](https://github.com/cerner/cerner_splunk/blob/stable/docs/vagrant.md) doc.

File: Gemfile [hunk 3]
```diff
@@ -3,16 +3,19 @@ source 'https://rubygems.org'
...
+chef_version = '= 14.5.27'
 
 gem 'berkshelf'
 gem 'chef', chef_version
 gem 'chef-sugar'
 gem 'chef-vault', chef_vault_version
 gem 'chefspec', chefspec_version
 gem 'foodcritic', foodcritic_version
+gem 'kitchen-inspec', kitchen_inspec_version
+gem 'kitchen-vagrant', kitchen_vagrant_version
```
Comment (gravesb) [inline Gemfile]: minor, why add these?  test kitchen isn't used in this cookbook (no .kitchen.yml file, no test/integration directory).

Comment (som3guy) [reply Gemfile]: I removed this in the new commit, sorry I was prepping to use test-kitchen for converging before finding another solution.

File: attributes/_user_management.rb [hunk 4]
```diff
@@ -4,6 +4,7 @@
 if node['platform_family'] == 'windows'
   default['splunk']['user'] = 'SYSTEM'
   default['splunk']['group'] = 'SYSTEM'
+  default['splunk']['password'] = 'splunk'
```
Comment (acharlieh) [inline attributes/_user_management.rb]: Guaranteed this isn't the SYSTEM password... so it shouldn't be set this way,

But also should we run Splunk as a different user on Windows?

File: Vagrantfile [hunk 5]
```diff
@@ -76,7 +76,7 @@ end
 
 Vagrant.configure('2') do |config|
   config.vm.box = 'bento/centos-6.7'
-  config.ohai.primary_nic = 'eth1'
+  config.ohai2.primary_nic = 'eth1'
```
Comment (acharlieh) [inline Vagrantfile]: Log an issue so that we handle multiple NICs properly. It'd be nice to not have to rely on this...

It would also be nice to manage address registration for indexers as well...

But that's scope creep.

File: spec/unit/recipes/_install_spec.rb [hunk 6]
```diff
@@ -120,12 +119,11 @@
 
     context 'when platform is rhel' do
       let(:platform) { 'centos' }
-      let(:platform_version) { '6.6' }
+      let(:platform_version) { '6.9' }
 
       it 'installs downloaded splunk package and notifies splunk-first-run' do
         expected_attrs = {
-          source: splunk_filepath,
-          provider: Chef::Provider::Package::Rpm
```
Comment (acharlieh) [inline spec/unit/recipes/_install_spec.rb]: Are we getting rid of provider object specifications? I don't see a change to do that, but we're getting rid of these attributes...?

**Comment by gravesb** [comment]
@som3guy Can you please respond to @acharlieh's comments/questions?

**Comment by som3guy** [comment]
Sorry I have been working through getting spec tests working with the data bag stuff I had to add for the windows password stuff.

File: recipes/_install.rb [hunk 7]
```diff
@@ -72,6 +72,7 @@ def nsp
   command "#{node['splunk']['cmd']} help commands --accept-license --answer-yes --no-prompt"
   user node['splunk']['user']
   group node['splunk']['group']
+  password node['splunk']['password'] if platform_family?('windows')
```
[the remark this answers is not in the document]
Comment (som3guy) [reply recipes/_install.rb]: Removing this in my next commit in favor of an encrypted_data_bag method.

File: Vagrantfile (hunk 5 above)
Comment (som3guy) [reply Vagrantfile]: Writing a new library to handle the work that used to be done by the ohai vagrant plugin.

File: spec/unit/recipes/_install_spec.rb (hunk 6 above)
Comment (som3guy) [reply spec/unit/recipes/_install_spec.rb]: Still going to be removed, Providers usage changed in [FC109](http://www.foodcritic.io/#FC109)

I will have changes reflecting this in my next commit.

File: attributes/_user_management.rb (hunk 4 above)
Comment (som3guy) [reply attributes/_user_management.rb]: I have changed it to be something a consumer can store in an encrypted data bag.

File: libraries/ipaddress.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# coding: UTF-8
+
+# Cookbook Name:: cerner_splunk
+# File Name:: ipaddress.rb
+
+# coding: UTF-8
+
+# Cookbook Name:: cerner_splunk
+# File Name:: alerts.rb
+
```
Comment (acharlieh) [inline libraries/ipaddress.rb]: copypasta Lines 6-10

File: libraries/ipaddress.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+require 'json'
+
+module CernerSplunk
+  # Module contains functions to configure ipaddresses in a Splunk system
+  module Ipaddress
```
Comment (acharlieh) [inline libraries/ipaddress.rb]: nit IPAddress

File: libraries/ipaddress.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    def self.default_ipaddress(node)
```
Comment (acharlieh) [inline libraries/ipaddress.rb]: management ip?

File: metadata.rb [hunk 11]
```diff
@@ -4,17 +4,15 @@
...
 source_url       'https://github.com/cerner/cerner_splunk'
 issues_url       'https://github.com/cerner/cerner_splunk/issues'
 
-chef_version     '~> 12.4' if respond_to?(:chef_version)
+chef_version     '>= 12.7', '< 15' if respond_to?(:chef_version)
 
 depends          'chef-vault', '~> 3.0'
-depends          'ulimit', '~> 0.3'
-depends          'xml', '~> 1.2'
-
```
Comment (acharlieh) [inline metadata.rb]: Style: keep this separation between cookbook dependencies and supported platform metadata....

File: attributes/_install.rb [hunk 12]
```diff
@@ -39,13 +39,3 @@
       '-x86-release.msi'
     end
   end
-
-default['splunk']['package']['provider'] =
```
Comment (acharlieh) [inline attributes/_install.rb]: Don't forget to remove this from documentation

File: .travis.yml [hunk 13]
```diff
@@ -5,7 +5,7 @@ language: ruby
 cache: bundler
 sudo: false
 rvm:
-- 2.3.1
+- 2.5.1
```
Comment (acharlieh) [inline .travis.yml]: Should we do both a chef 12 and a chef 14 run?

File: Vagrantfile [hunk 14]
```diff
@@ -44,7 +44,7 @@ fail 'Non-unique hostnames' if @network.collect { |_, v| v[:hostname] }.uniq!
 fail 'Non-unique ports' if @network.collect { |_, v| v[:ports].keys }.flat_map { |v| v }.uniq!
 
 def default_omnibus(config)
-  config.omnibus.chef_version = '12'
+  config.omnibus.chef_version = '14.5.27'
```
Comment (acharlieh) [inline Vagrantfile]: why so specific?

Comment (nmattam) [reply Vagrantfile]: On line 355 we have defined the chef-client URL for windows forwarder. We will need to update that as well.
https://github.com/cerner/cerner_splunk/blob/00aa73d9b1f72969dd9963daeff3349a760cbfe5/Vagrantfile#L354-L355

File: .travis.yml (hunk 13 above)
Comment (som3guy) [reply .travis.yml]: Why would we want to support an unsupported version of Chef with these new updates?
I am thinking no, but if you know a good reason then certainly.

File: Vagrantfile (hunk 14 above)
Comment (som3guy) [reply Vagrantfile]: I had to be specific, the vagrant runs were failing when I only specified `14`.

File: .travis.yml (hunk 13 above)
Comment (acharlieh) [reply .travis.yml]: If our cookbook supports both chef 12 and 14, then it's not a major version number bump, and we make it easier on consumers who have to do both Chef upgrades and cookbook updates, as opposed to forcing a lock step of upgrade chef, then upgrade cookbook.

File: Vagrantfile (hunk 14 above)
Comment (som3guy) [reply Vagrantfile]: @nmattam fixed in 3ee2707

Comment (acharlieh) [reply Vagrantfile]: What was the reason for the failure? That seems an issue that needs to be addressed if we cannot just pull the latest chef 14 client...

File: attributes/_install.rb (hunk 12 above)
Comment (som3guy) [reply attributes/_install.rb]: removed in 3ee2707

File: docs/attributes.md [hunk 15]
```diff
@@ -40,11 +39,11 @@ Configurable (with defaults)
...
-* `node['splunk']['mgmt_host']` - The host other SHC members use when connecting to the current node. You probably want a wrapper cookbook to override this. (`node['ipaddress']`)
 * `node['splunk']['heavy_forwarder']['use_license_uri']` - Set this attribute to `true` to point the Heavy Forwarder to the license master. (`false`)
 * `node['splunk']['apps']` - An [apps hash](databags.md#apps-hash) of apps to configure locally. (Does not support downloading apps ... yet...)
 * `node['splunk']['data_bag_secret']` - The location of the shared secret file if your encrypted data bags are encrypted via shared secret rather than chef-vault. If this is not specified, and the encrypted data bags are using shared secret encryption then chef looks for a secret at the path specified by the encrypted_data_bag_secret setting in the client.rb file.
 * `node['splunk']['forwarder_site']` - Set this attribute to configure site awareness for your forwarders.(`site0`)
+* `node['splunk']['mgmt_interface']` - You can configure this to what you want your default network interface to be. Default is `node['default_interface']`
```
Comment (gravesb) [inline docs/attributes.md]: minor
```suggestion
* `node['splunk']['mgmt_interface']` - You can configure this to what you want your default network interface to be. Default is `node['network']['default_interface']`
```

File: metadata.rb [hunk 16]
```diff
@@ -4,17 +4,16 @@
...
 description      'Installs/Configures Splunk Servers and Forwarders'
 long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
-version          '2.21.1'
+version          '2.21.2'
 
 source_url       'https://github.com/cerner/cerner_splunk'
 issues_url       'https://github.com/cerner/cerner_splunk/issues'
 
-chef_version     '~> 12.4' if respond_to?(:chef_version)
+chef_version     '>= 12.7', '< 15' if respond_to?(:chef_version)
```
Comment (gravesb) [inline metadata.rb]: minor, chef_version was [introduced in 12.6](https://github.com/chef/chef/commit/98a1793e95e7f89b5206883528a0d00c209838f0#diff-e7f90eec970c3a67a738e9000abf630c) so since we're requiring 12.7 or greater we don't need the `if respond_to?()` check anymore.

File: recipes/_install.rb [hunk 17]
```diff
@@ -55,23 +55,39 @@ def nsp
   only_if(&manifest_missing)
 end
 
-package node['splunk']['package']['base_name'] do
-  source splunk_file
-  version "#{node['splunk']['package']['version']}-#{node['splunk']['package']['build']}"
-  provider node['splunk']['package']['provider']
-  only_if(&manifest_missing)
-  if platform_family?('windows')
+if platform_family? 'rhel', 'fedora', 'amazon'
```
Comment (gravesb) [inline recipes/_install.rb]: I get why windows is separate, but is there a reason we can't just use the package resource and let chef determine the proper provider to use automatically for the others?

Comment (acharlieh) [reply recipes/_install.rb]: Because these aren't the automatically selected providers. On [EL Chef selects yum](https://github.com/chef/chef/blob/v14.6.35/lib/chef/provider/package/yum.rb#L40) (which might be ok... not entirely sure, we were manually picking the RPM provider earlier), on Debian flavors [chef selects apt-get](https://github.com/chef/chef/blob/v14.6.35/lib/chef/provider/package/apt.rb#L28) (which [cannot use a local source](https://github.com/chef/chef/blob/v14.6.35/lib/chef/provider/package/apt.rb#L45-L48) )

File: Vagrantfile (hunk 14 above)
Comment (acharlieh) [reply Vagrantfile]: We should figure out where that message is coming from since invoking the chef install script directly from a container seems to taking `12` and `14` and resolving the latest version of 12 and 14 just fine: 

```
[root@84cf013a9463 /]# curl https://www.chef.io/chef/install.sh -O; sh install.sh -v 14
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 23082  100 23082    0     0  77050      0 --:--:-- --:--:-- --:--:--  256k
el 6 x86_64
Getting information for chef stable 14 for el...
downloading https://omnitruck-direct.chef.io/stable/chef/metadata?v=14&p=el&pv=6&m=x86_64
  to file /tmp/install.sh.192/metadata.txt
trying curl...
sha1	dc1cfb5bee1e6f0df01c80b55ff7cd2fa22abc7e
sha256	219167dc31fcb1cef1fdd669403ccd91066ca23cf75e74da12e144c4b816fda6
url	https://packages.chef.io/files/stable/chef/14.5.33/el/6/chef-14.5.33-1.el6.x86_64.rpm
version	14.5.33
downloaded metadata file looks valid...
downloading https://packages.chef.io/files/stable/chef/14.5.33/el/6/chef-14.5.33-1.el6.x86_64.rpm
  to file /tmp/install.sh.192/chef-14.5.33-1.el6.x86_64.rpm
trying curl...
^C

[root@84cf013a9463 /]# curl https://www.chef.io/chef/install.sh -O; sh install.sh -v 12
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 23082  100 23082    0     0  84480      0 --:--:-- --:--:-- --:--:--  262k
el 6 x86_64
Getting information for chef stable 12 for el...
downloading https://omnitruck-direct.chef.io/stable/chef/metadata?v=12&p=el&pv=6&m=x86_64
  to file /tmp/install.sh.225/metadata.txt
trying curl...
sha1	ecb40884ed9fc7064eb04930c0695757e77aab8f
sha256	cf3c41d029c24ede942fa9c756990407b0396197ec6422df852028493bd6e3ef
url	https://packages.chef.io/files/stable/chef/12.22.5/el/6/chef-12.22.5-1.el6.x86_64.rpm
version	12.22.5
downloaded metadata file looks valid...
downloading https://packages.chef.io/files/stable/chef/12.22.5/el/6/chef-12.22.5-1.el6.x86_64.rpm
  to file /tmp/install.sh.225/chef-12.22.5-1.el6.x86_64.rpm
trying curl...
^C
```

Comment (som3guy) [reply Vagrantfile]: It likes a integer not a string i guess.

File: docs/attributes.md (hunk 15 above)
Comment (som3guy) [reply docs/attributes.md]: fixed in: 
54f10dc

File: metadata.rb (hunk 16 above)
Comment (som3guy) [reply metadata.rb]: removed in: accc160

File: Vagrantfile (hunk 14 above)
Comment (som3guy) [reply Vagrantfile]: fixed in: c9f3f7d

File: docs/attributes.md [hunk 18]
```diff
@@ -40,11 +39,11 @@ Configurable (with defaults)
...
-* `node['splunk']['mgmt_host']` - The host other SHC members use when connecting to the current node. You probably want a wrapper cookbook to override this. (`node['ipaddress']`)
 * `node['splunk']['heavy_forwarder']['use_license_uri']` - Set this attribute to `true` to point the Heavy Forwarder to the license master. (`false`)
 * `node['splunk']['apps']` - An [apps hash](databags.md#apps-hash) of apps to configure locally. (Does not support downloading apps ... yet...)
 * `node['splunk']['data_bag_secret']` - The location of the shared secret file if your encrypted data bags are encrypted via shared secret rather than chef-vault. If this is not specified, and the encrypted data bags are using shared secret encryption then chef looks for a secret at the path specified by the encrypted_data_bag_secret setting in the client.rb file.
 * `node['splunk']['forwarder_site']` - Set this attribute to configure site awareness for your forwarders.(`site0`)
+* `node['splunk']['mgmt_interface']` - You can configure this to what you want your default network interface to be. Default is `node['network']['default_interface']`
```
Comment (gravesb) [inline docs/attributes.md]: Similar to the config this is replacing, I think it would be helpful to indicate it is only relevant for search head clustering. Something along the lines of: "The network interface this node should use for communicating with other members of a SHC. (`node['network']['default_interface']`)"

File: recipes/_install.rb [hunk 19]
```diff
@@ -55,23 +55,39 @@ def nsp
...
     # installing as the system user by default as Splunk has difficulties with being a limited user
     options %(AGREETOLICENSE=Yes SERVICESTARTTYPE=auto LAUNCHSPLUNK=0 INSTALLDIR="#{node['splunk']['home'].tr('/', '\\')}")
   end
+else
+  fail 'unsupported platform'
 end
 
 include_recipe 'cerner_splunk::_configure_secret'
 
+windows_password = CernerSplunk::DataBag.load(node['splunk']['windows_password'], secret: node['splunk']['data_bag_secret'])
```
Comment (gravesb) [inline recipes/_install.rb]: `node['splunk']['windows_password']` needs to be added to the attributes doc.

Comment (som3guy) [reply recipes/_install.rb]: added in ebac2d9

File: docs/attributes.md (hunk 18 above)
Comment (som3guy) [reply docs/attributes.md]: updated in 018f21e

File: .travis.yml (hunk 13 above)
Comment (som3guy) [reply .travis.yml]: I am not seeing how to do this, I believe the version is hardcoded in the [Gemfile](https://github.com/cerner/cerner_splunk/blob/stable/Gemfile#L10)

Comment (som3guy) [reply .travis.yml]: Added the updates here: 312b003

File: recipes/_configure_server.rb [hunk 20]
```diff
@@ -178,7 +179,7 @@
   server_stanzas['shclustering']['pass4SymmKey'] = CernerSplunk::ConfTemplate.compose encrypt_password, CernerSplunk::ConfTemplate::Value.constant(value: pass) if pass
   server_stanzas['shclustering']['conf_deploy_fetch_url'] = deployer_uri
   server_stanzas['shclustering']['disabled'] = 0
-  server_stanzas['shclustering']['mgmt_uri'] = "https://#{node['splunk']['mgmt_host']}:8089"
+  server_stanzas['shclustering']['mgmt_uri'] = "https://#{management_ipaddress}:8089"
```
Comment (gravesb) [inline recipes/_configure_server.rb]: I just realized this is a non-passive change that I believe would require a major version bump of the cookbook. It will break existing search head clusters without additional configuration changes on the consumers end. What was the reason behind this change from the configurable management host to pulling the ip address from the ohai data?

Comment (som3guy) [reply recipes/_configure_server.rb]: Not sure that is correct, `node['splunk']['mgmt_host']` = `node[ipaddress]` which I believe was the ipaddress of the default_interface.

By default `management_ipaddress` is the same thing, but this allowed us to not have to use the vagrant ohai plugin. 

In the future would allow us specify an interface to use for something specific. So the boxes with many interfaces each interface could be used for something specific rather than one interface doing all of the work.

At least I think that was all of the reasons behind it, @acharlieh could correct me if I am wrong.

Comment (som3guy) [reply recipes/_configure_server.rb]: Understood, so which path should I go down?
Removed the recent work I did to move towards making testing work for both chef 12 and 14 or find a way to make it not a breaking change?

File: Vagrantfile [hunk 21]
```diff
@@ -3,7 +3,7 @@
 
 Vagrant.require_version '>= 1.4.1'
 
-%w[vagrant-ohai2 vagrant-omnibus].each do |plugin|
```
Comment (nmattam) [inline Vagrantfile]:
```
==> c2_boot1: Running chef-client...
==> c2_boot1: [2018-10-24T21:40:58+00:00] FATAL: Configuration error NoMethodError: undefined method `<<' for nil:NilClass
==> c2_boot1: [2018-10-24T21:40:58+00:00] FATAL:   /tmp/vagrant-chef/client.rb:32:in `from_string'
==> c2_boot1: [2018-10-24T21:40:58+00:00] FATAL: Aborting due to error in '/tmp/vagrant-chef/client.rb'
```

@som3guy  I keep running into this on your branch and I believe you had the fix for this in vagrant-ohai2. Right?
https://github.com/avishai-ish-shalom/vagrant-ohai/pull/21/files

Comment (gravesb) [reply Vagrantfile]: There's no way to enable or disable vagrant plugins.  If you have it installed, it will get used.  I had to uninstall the vagrant-ohai plugin to get the vagrant nodes to come up on this branch. With the latest changes, the vagrant-ohai2 plugin is also unnecessary.

Comment (nmattam) [reply Vagrantfile]: Yep.. uninstalling fixed it.

File: docs/attributes.md [hunk 22]
```diff
@@ -40,12 +39,12 @@ Configurable (with defaults)
 * `node['splunk']['config']['ui_prefs']` - Hash of stanzas used to configure [ui-prefs.conf][] on the search head in a clustered configuration or a standalone instance.
 * `node['splunk']['config']['assumed_index']` - Name of the index to which data is forwarded to by default, when the index is not configured for the input.(`main`)
 * `node['splunk']['bootstrap_shc_member']` - Set this attribute to `true` to bootstrap a member to the Search Head Cluster (SHC). (`false`)
-* `node['splunk']['mgmt_host']` - The host other SHC members use when connecting to the current node. You probably want a wrapper cookbook to override this. (`node['ipaddress']`)
```
Comment (nmattam) [inline docs/attributes.md]: With the recent changes, we need this with nil as the default value.

Comment (som3guy) [reply docs/attributes.md]: fixed in 367cc13

File: docs/attributes.md [hunk 23]
```diff
@@ -40,12 +39,13 @@ Configurable (with defaults)
...
 * `node['splunk']['heavy_forwarder']['use_license_uri']` - Set this attribute to `true` to point the Heavy Forwarder to the license master. (`false`)
 * `node['splunk']['apps']` - An [apps hash](databags.md#apps-hash) of apps to configure locally. (Does not support downloading apps ... yet...)
 * `node['splunk']['data_bag_secret']` - The location of the shared secret file if your encrypted data bags are encrypted via shared secret rather than chef-vault. If this is not specified, and the encrypted data bags are using shared secret encryption then chef looks for a secret at the path specified by the encrypted_data_bag_secret setting in the client.rb file.
 * `node['splunk']['forwarder_site']` - Set this attribute to configure site awareness for your forwarders.(`site0`)
-
+* `node['splunk']['mgmt_host']` - The host other SHC members use when connecting to the current node. You probably want a wrapper cookbook to override this. By default `node['splunk']['mgmt_interface']` is now used, but to support existing configurations this attribute is still available and takes precidence when set. (`nil`)
```
Comment (gravesb) [inline docs/attributes.md]: minor precedence misspelled

**Comment by som3guy** [comment]
I think I have addressed all concerns and nit's if I could get another look on this providing the tests pass.

**Comment by som3guy** [comment]
@nmattam Could you take another look at this PR
